<a href="https://colab.research.google.com/github/j-claudinei-f/j-claudinei-f/blob/main/Recorr%C3%AAncias_ordem1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Matemática Discreta**

José Claudinei Ferreira

Universidade Federal de Alfenas

#**Relações de recorrência lineares de ordem 1**

[Link](https://github.com/j-claudinei-f/j-claudinei-f/blob/main/SomasPolinomiais.ipynb) para somas polinomiais e exponenciais, para resolução de relações de reconrência da forma
$$a_{n+1}=a_n+f(n+1),$$ com $f(n)$ e $a_1$ conhecidos.

#**Relações de recorrência lineares de ordem 1**

Agora estamos interessados nas resolução de relações de reconrência da forma
$$a_{n+1}=g(n)a_n+h(n),$$ com $g(n)$, $h(n)$ e $a_1$ conhecidos.

O método de resolução é o seguinte (Teorema 4.1 do livro texto):

1. Considere uma sequência $b_n$, solução não nula da recorrência <font color=blue>$$b_{n+1}=g(n)b_n.$$

2. Defina $y_n$ como uma sequência, tal que <font color=green> $$b_ny_n=a_n.$$

3. Segue que $$\begin{cases}\color{green}{b_{n+1}y_{n+1}}&=&\color{green}{a_{n+1}}\\&=&g(n)a_n+h(n)\\&=&g(n)\color{green}{b_n y_{n}}+h(n)\\&=&\color{blue}{g(n)b_n }y_{n}+h(n)\\&=&\color{blue}{b_{n+1}}y_n+h(n)\end{cases}$$

Considerando o início e o final da última exeperssão, temos que $$y_{n+1}=y_n+\frac{h(n)}{g(n)b_n},\qquad y_1b_1=a_1,$$  que é uma expressão mais simples de manipular, usando somatório.

#**Exemplo**

 Considere a relação de recorrência $$a_{n+1}=3 a_n+2,$$ com $a_1=2$, que vem de uma forma de calcular o número mínimos de movimentos na torre de Hanói (com a regra de sempre passar pela haste do meio). 

Calcule $a_{10}$ e determine uma expessão para $a_n$ sem recorrência.

**Uma parte da solução:**

Para calcular $a_{10}$ basta calcularmos $a_2,\,a_3,\,\ldots,\,a_9$ e $a_{10}$, usando a recorrência.

In [ ]:
a=[2]

for i in range(1,10):
  a=a+[3*a[i-1]+2]

a

[2, 8, 26, 80, 242, 728, 2186, 6560, 19682, 59048]

Isso nos dá $a_{10}=59048$.

**Uma solução geral**

Uma forma de resolver esse problema é resolver antes a recorrência
$$b_{n+1}=3b_n,$$ que tem $b_n=3^{n}$ como solução, por exemplo. 

Agora obtemos uma relação de recorrência $$y_{n+1}=y_n+\frac{2}{3^{n}\times 3},\qquad 3^{1}y_1=a_1=2.$$

Devemos ter $a_n=3^ny_n$.

Mas, $$y_2=\frac{2}{3}+\frac{2}{3}\frac{1}{3^1}=\frac{2}{3}\left(1+\frac{1}{3^1}\right)$$
$$y_3=\frac{2}{3}\left(1+\frac{1}{3^1}+\frac{1}{3^2}\right)$$
e podemos ver que 
$$y_n=\frac{2}{3}\left(1+\frac{1}{3^1}+\frac{1}{3^2}+\cdots+\frac{1}{3^{n-1}}\right)=\frac{2}{3}\left(\frac{\frac{1}{3^{n}}-1}{\frac{1}{3}-1}\right).$$

Por fim, $$a_n=3^n\frac{2}{3}\left(\frac{\frac{1}{3^{n}}-1}{\frac{1}{3}-1}\right)=3^n-1.$$

In [ ]:
3**10-1   # 3^10-1=a(10)

59048

#**Uma forma para somar produto de expressões**

Quando formas considerar relações de rerrência da forma 
$$a_{n+1}=a_n+f(n+1),$$
podemos começar com uma certo $a_1$ e obter $a_2=a_1+f(2)=(a_1-f(1))+f(1)+f(2)$, ...,$$$a_n=a_1+f(2)=(a_1-f(1))+f(1)+f(2)+\cdots f(n),$$
como temos feito antes.

Já vimos como calcular a soma, de forma reduzida, para os casos em que $$f(n)=\sum_{i=1}^pc_ii^m$$ é um polinômio ($m\in\mathbb{N}$) e no caso em que $$f(n)=\sum_{i=1}^pc_ir_i^i$$ é uma soma de exponenciais.

Segue que sabemos manipular o somatório no caso em que $f(n)$ é uma soma de polinômios com exeponenciais (da forma amnecionada).


Entretanto, há casos em que $$f(n)=g(n)h(n)$$ é um produto e podemos usar a seguinte ideia para tentar manipular o somatório 
$$\sum_{i=1}^nf(i)=\sum_{i=1}^ng(i)h(i).$$